src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -51,46 +52,121 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
+            Debug.Assert(elementNames.IsDefaultOrEmpty || elementTypes.Length == elementNames.Length);
+            Debug.Assert(elementTypes.Length > 1 && elementTypes.Length < 8); // PROTOTYPE
 
-            switch (elementTypes.Length)
+            var tupleType = binder.GetWellKnownType(EnumType(elementTypes.Length), diagnostics, syntax);
+            NamedTypeSymbol underlyingType = tupleType.Construct(elementTypes);
+
+            if (elementTypes.Length > 1 && elementTypes.Length < 8)
             {
-                case 2:
-                    {
-                        var tupleType = binder.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_ValueTuple_T1_T2, diagnostics, syntax);
-                        underlyingType = tupleType.Construct(elementTypes);
-
-                        var underlyingField1 = (FieldSymbol) Binder.GetWellKnownTypeMember(binder.Compilation, WellKnownMember.System_Runtime_CompilerServices_ValueTuple_T1_T2__Item1, diagnostics, syntax: syntax);
-                        var underlyingField2 = (FieldSymbol) Binder.GetWellKnownTypeMember(binder.Compilation, WellKnownMember.System_Runtime_CompilerServices_ValueTuple_T1_T2__Item2, diagnostics, syntax: syntax);
-
-                        fields = ImmutableArray.Create(
-                                    new TupleFieldSymbol(elementNames.IsDefault ? "Item1" : elementNames[0],
-                                        this,
-                                        elementTypes[0],
-                                        underlyingField1?.AsMember(underlyingType)),
-                                    new TupleFieldSymbol(elementNames.IsDefault ? "Item2" : elementNames[1],
-                                        this,
-                                        elementTypes[1],
-                                        underlyingField2?.AsMember(underlyingType))
-                                );
-
-                        return underlyingType;
-                    }
+                var fieldSymbols = ArrayBuilder<TupleFieldSymbol>.GetInstance(elementTypes.Length);
 
-                default:
-                    {
-                        // TODO: VS if this eventually still stays reachable, need to make some error type symbol
-                        var tupleType = binder.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_ValueTuple_T1_T2, diagnostics, syntax);
-                        underlyingType = tupleType.Construct(elementTypes);
-                        break;
-                    }
+                for (int i = 0; i < elementTypes.Length; i++)
+                {
+                    var wellKnownTypeMember = EnumTypeMember(elementTypes.Length, i);
+                    var underlyingField = (FieldSymbol)Binder.GetWellKnownTypeMember(binder.Compilation, wellKnownTypeMember, diagnostics, syntax: syntax);
+                    var field = new TupleFieldSymbol(elementNames.IsDefaultOrEmpty ? UnderlyingMemberName(i) : elementNames[i],
```

Comment:
@cston, Do you mean that I should only check for empty or default, but not both?